### ISSUE TITLE
Handling firewalls

### ISSUE THREAD

**Issue by Nadahar** [UniversalMediaServer/UniversalMediaServer#686]
@UniversalMediaServer/developers @SubJunk 
It seems to me like UMS isn't taking care of firewalls and simply assumes that everybody will turn them off. It can be quite tricky to deal with, as there are a number of different software firewalls available and it's hard to test incoming access without cooperating with an external host. Most OS'es these days also comes with firewalls enabled by default, and not-so-technical users will be scared to turn them off even if they are protected by a router and NAT.

I do think that more could be done though. The very least we should do is to inform the users of what incoming ports need to be open for UMS to work. HTTP wise that's pretty simple, just two TCP ports (one for the DLNA server and one for the web interface)? On the uPnP/UDP side I really don't know what is needed, uPnP is a messy standard network wise with it's broadcasting and whatnot. It shouldn't be to hard for us to figure out what was needed though (isn't UDP port 5000 involved?), and we should probably show this information in the GUI dynamicly so that it reflects the current configuration.

What would be really nice is if it would be possible to contact a renderer and request for it to contact UMS back (instead of replying in the estabilished channel) and in that way be able to test that it could reach UMS. I don't know if there are any such possibilities in DLNA or uPnP though.

We should also offer to open the necessary ports at the very least on Windows firewall, from the installer. I don't know if the install processes on the other OS'es would allow something like that, Linux is hopeless in that there's no standard for anything, and OS X is just foreign to me.

What do you guys think?

**Comment by Nadahar** [comment]
Here's some [more information] (http://universalmediaserver.com/forum/viewtopic.php?f=9&t=5778&p=21813#p21813) about the subject. Firewall issues are by far the most common support thread topic, and I'd be willing to bet that a lot of people simply abandon UMS before going to any forums because they can't get it to work for this reason.

**Comment by WolfganP** [comment]
I agree with @Nahadar, most of the "UMS doesn't detect my renderer" or viceversa is one of the most repeated subjects at the forums (mostly in Windows) and firewall/AV related. 
Isn't there any way for UMS to include a firewall blocking test routine at first start that may alert users? And/or a settings button that may open the Firewall Rules window for the users to check against a FAQ article or specific forum post?

**Comment by Nadahar** [comment]
@WolfganP It's not easy to check for incoming access to the computer you're actually on. The only way I can think of is if there's a DLNA/UPnP command we could send to a renderer (if it were detected that is) that would result in a predictable reply that we could use to determine that traffic was being blocked if it didn't arrive. The problem is that nothing is predictable in DLNA/UPnP, nobody "follows the rules" and the "rules" themselves are a mess. You need to cooperate with another device to check for incoming traffic, and that's hard to do if incoming traffic is blocked.

Actually, I don't understand how UMS will ever see any renderers if incoming access is blocked, and the reported cases of this must mean that some ports are blocked and some are open as far as I can understand. 

When it comes to checking if Windows firewall is active, it probably needs admin permissions and I'm pretty sure that doing such a thing is close to impossible from Java. Java's interaction with native libraries is terrible, unless you want to dig really deep into it with JNI. I have no idea how to use JNI, it looks to me like a terrible mess.

Creating some kind of information message that would pop up after installing would probably annoy all those users that do know what they are doing, and another big chunk would probably complain that we recommend turning off the firewall. 

Some programs like e.g uTorrent has firewall checks, but the way they do it is by contacting a server on the internet that then attempts to contact the local uTorrent application on the same IP address. The problem with UMS is that we can't use a server on the internet to check, since it's not access from the internet that's required. Normally the router will block all incoming traffic and that's fine for UMS, we only need incoming access from the local network. If testing from a internet server, there would be now way of testing the "inner" firewall that is running on the UMS computer unless the "outer" (the router) was open.

I honestly don't know how to handle this, but a start would be to get a full overview over what's needed ourselves. Then we could create a document describing it in detail at least, and maybe create the necessary access in the installer. I bet it's a nightmare though, there are many different versions of the Windows firewall in the different Windows versions, so I bet we'd have to cater for that.

**Comment by WolfganP** [comment]
I agree on start documenting the topic or maybe create an specific KB in the forum so we can quickly point the users with issues to it.
On the automation side, maybe there's something to explore on the UPNP side of things... ie https://github.com/kaklakariada/portmapper

**Comment by Nadahar** [comment]
Portforwarding via UPnP is something entirelt different. UPnP is an "umbrella" standard, containing several standards. UMS use the UPnP AV while UPnP capable routers use the Internet Gateway Device Protocol.

**Comment by sainokawara** [comment]
Nadahar, I always wondered why I needed a GitHub Logon.  As a retired techical team-leader for the design & development of Operating Systems, Database & Communication Managers [on mainframes but don't tell anyone] who blamed all problems on Testers who couldn't tell me what they did that lead to a failure and users who did not RTF manual, I am uniquely unqualified for this task!  And, my writing skills are nonexistent as well!

However, I am absolutely determined to see something done on this topic. Give me a few weeks to see what I can put together on this topic.  I have downloaded the UPnP spec and am agast at some of the strange things that were done.  Did you know that UPnP devices are allowed to create IP addresses in the range 169.254.0.0/16?  When I ran WireShark on UMS, I wondered from where these strange IPs came.  And, you don't want to know all of the things they do with 239.255.255.250:1900.  Unfortunately, there is also a lot of "legacy" remnants like Port 5000 that may or may not still be used because of the "backward compatibility" statement.  I don't know if a Firewall rule should be created for it or not?

I have also downloaded the UMS code but will have to learn enough Java as dissect, or is it bisect, it.  I may be old & retired but I am, at least hope, still capabile of learning new things.  I also found "Cling" although I do not know enough about it yet to even know what it is or how it is used.  I was a little concerned when the latest date I could find on the website was 2013.  I hope it is supported by a consortium that will be around for a long time?

Finally, the various pointers you gave me to websites and the like were very helpful.  If you know of any others, I will investigate them.  You can email me at sainokawara.sisyphus@gmail.com and I, hopefully, have subscribed to the relevant Forum Posts.  Currently, my Test environment is limited to 64-bit Windows 10, Kubuntu running under Hyper-V, a Samsung TV and an Android Tablet where I was able to get the Kodi APP working with UMS.

And,WolganP, should I recognize that "creature"?  I wonder if I should come up with something?

**Comment by sainokawara** [comment]
Do you happen to recall whether people who had Firewall "problems" were trying to run as a Windows Service?  The reason I ask is because when Java is installed, two Firewall "rules" are created basically allowing ALL TCP Ports and ALL UDP Ports through for "C:\Program Files [(x86]\Java\jre1.8.0_nn\bin\javaw.exe".  Unless something needs ICMP or IGMP, this should allow everything under the sun through for any Java Application including UMS.

And, I have finally discovered a selective "rule" that allows UMS to run as a Windows Service the "old-fashion way": trial-and-error.  All that is needed is to allow TCP Ports 5001, 9001 through to "C:\Program Files [(x86)]\Java\jre1.8.0_nn\bin\java.exe"!  However, unlike Sun/Oracle, I don't think UMS is going to get away with just allowing EVERYTHING and ANYTHING through like Java.  So, I will keep digging to see whether there is anything else blocked in UPnP/SSDP.

However, I discovered that it is extremely important to allow one or two MINUTES to pass before you assume that UMS startup failed or a UPnP device is unable to find/connect to UMS!  There is a very simple but ugly explanation due to the way "Auto-IP"s are allocated.  And, I see some created on my system although I do not know why.  All my systems/devices have static IPs so UPnP should have just used the IPs returned by DHCP which should have been virtually instantaneous!  So, I have no idea of what/why "Auto-IP" are even being created?

**Comment by Nadahar** [comment]
@sainokawara First of all, I'd like to say that we really appreciate that you'll take on trying to figure out this. It's badly needed, that's for sure. 

Regarding cling, I don't know if you found the right one. Here is the [cling web site] (http://4thline.org/projects/cling/), and the latest release is currently from 9/12-2015 - so it's in no way stale.

About things like port 5000 it's hard to say. If there are any circumstances were it will be used, it should be opened. The risks of opening specific (not being standard ports for typically attack targets) ports on a local firewall is so minimal that I don't think we should hold back there. In the end, if the content on the computer in question is that sensitive, they probably shouldn't install UMS in any case, it is after all a media delivery software using an unauthenticated protocol. It's more important that if these rules are followed, everything will work properly as I see it. If the port is never used in our context though, there's no reason to open it.

Regarding Java's firewall rules, remember that this is only outgoing connections. Incoming connections are blocked by default, so the renderer won't see UMS and can't respond to multicast broadcasts. While TCP has a state, UDP is stateless. I'm sure you know this, but it means that when an outbound TCP connection is estabilished, replies are allowed to travel back within the estabilished channel without an inbound rule. TCP connections only need to be allowed access when the connection is estabilished so to say. UDP on the other hand has no concept of a channel or connection, so an UDP reply will need inbound access even to be able to respond.

When UMS is running as a service, a wrapper is in use. I suspect, but haven't tested this, that ```wrapper.exe``` needs the same permissions as javaw has. I don't think we should trust that Java has applied it's rules either. We can note in the documentation that these rules are normally created by Java on Windows, but still include them. I see no problem allowing any outbound ports also for services, weather it's ```wrapper.exe``` or ```java.exe```. TCP ports 5001 and 9001 are inbound rules and they are unrelated to UPnP. 

TCP port 5001 is a http UMS runs to serve contents to the renderers. UPnP is used for discovery and device control, but the actual content delivery is done over HTTP. I suspect that UMS informs that it's available on TCP port 5001 in it's UPnP broadcast, so that renderers know how to request content.

TCP port 9001 is also a http server, but this one is for delivering UMS' web interface. It only needs opening if the web interface are to be available, if it's not in use no other functionality should be affected by this port being blocked.

Both TCP port 5001 and 9001 are configurable from the UMS configuration. Some people might choose to use different ports (e.g port 80 for the web interface), and some people run multiple instances of UMS simultaneously using different ports (for serving different renderers or different content). 5001 and 9001 are only the defaults, so the documentation must reflect this.

I don't know why these "Auto-IPs" are created, but I hope that has nothing to do with UMS. One explanation might be that most computers have multiple network interfaces even when only one "real" network interface is installed. There are all kind of virtual devices and whatnot installed by Windows and by various other software, and some of these can be set to use DHCP. If they have no real connectivity, they can end up "self assigning" IP addresses from the ```169.254.0.0/16``` range.

I'm a bit @skeptical about accepting that discovery can take several minutes. To me that indicates that something is blocked. Normally discovery should be near instant, as you can verify by turning off the firewall. As far as I know, broadcasts are made when looking for partners and then at regular intervals. I'm guessing that when UMS starts up, it will make a broadcast announcing that it's there (and the same when a renderer is turned on). But, if the renderers are unable to receive or reply to UMS' broadcast, UMS might not discover them until the renderer gets to it's next scheduled announcement. Even though it seems to work after a while, communication is then in reality crippled.

**Comment by sainokawara** [comment]
I believe I know the answer to most of the above questions but the acid test is seeing whether testing proves them correct.  However, first, I would like to understand how you would prefer to be informed.  I have no problem posting EVERYTHING here:  I run exhausting [of me] tests and take copious notes.  Would you like to see these? Or, I can just send summaries of my finds with references to how I arrived at that conclusion.  Although a software engineer by trade, I was trained in mathematics and [theoretical] physics and subscribe to the idea of "reproducible" experiments/tests. But, I do have one tidbit that I would like to share with you that I would prefer not to post.  Is there a way for me to do that on GitHub or did you write down my e-mail address?  Today, I want to run WireShare to verify the activity on TCP Ports 5000 and 2869.  The WireShark and WinPcap people did an outstanding job but I really hate sifting through all of the data.  By the way, I found a loose plug on an Ethernet cable between a switch and router that was causing intermittent problems.  Perhaps, that explains some of the very long events I was observing?  I hope so.

**Comment by sainokawara** [comment]
I completed WireShark traces for UMS running as a Windows Service for:
1) UMS "Clean" Startup
2) TV Connect/Use
3) Android Kodi APP Connect/Use
There were ***ZERO*** TCP/UDP Port 2869 or 5000 Records!
I will repeat for UMS running as a Java APP but I seriously doubt there will be any.

I do am not suggesting that we Firewall block these two Ports
but I also see no reason to create Firewall rules allowing them for UMS.

**Comment by sainokawara** [comment]
Although my knowledge of Java is limited to being able to spell it, there should be no need for additional Firewall rules when UMS is run as a Java Application.  My testing indicates that the
only Firewall rule that needs to be created when it is run as a Windows Service is:
"C:\Program Files ...\Java\jre1.8.0_nn\bin\java.exe" is for TCP 5001, 9001.
I did test "javaw.exe", "javaws.exe" and "wrapper.exe" but discovered that
there appears to be no need to create Firewall rules for them.  And, with this single rule,
I have no difficulty and there are no error messages when UMS runs as a Windows Service.

**Comment by sainokawara** [comment]
It's not even 8 o'clock in the morning and I discovered why UPnP AutoIPs were being created.  And, this may explain why discovery was taking several minutes for me.  I will let you know if this does not solve that issue.

I have multiple adapters on my system [although the problem still might occur on a single adapter system.  I will verify whether this is true or not, later.]  If the second adapter is connected to the network as any normal person would do, there would be no problem.  However, if the second adapter is reserved for Hyper-V, there are times when there is no IP address assigned to it [even in the DHCP server].  In this case, when UMS starts, it detects the second adapter but since there is no assigned IP, it has to create an Auto-IP for it!  The solution was so simple and obvious once I looked at the UMS logs after a nights sleep.

Which brings me to another point:  This document needs to contain the location for the UMS "debug log files"!  For Windows, it is "C:\ProgramData\UMS\".  But, when run as a Windows Service, there is also "C:\Program Files (x86)\Universal Media Server\wind32\service\".  You should also specify exactly which "debug log files" you might need to diagnose any problems.

**Comment by Nadahar** [comment]
Multiple adapters is quite common, many motherboards have multiple network adapters. I think it's quite common to only connect one of them, there would be no benefit using both unless they were used in teaming. Teaming requires a teaming-capable switch, which I suspect is rare in home networks, and thus they will often be left unconnected.

 

By default UMS will bind to all adapters, but I'd think that Windows would create the "Auto-IP" independent from UMS. Are you sure the IP isn't tied to the adapter before UMS starts? This can be seen easily by running "ipconfig /all" from the command line.

 

When it comes to the log file location it's not that simple.  The default location on Windows is "%programdata%\UMS". This can be changed however, de refer to this folder as UMS' profile folder. You can create multiple profiles and launch them via command line options. In addition to that, the location of %programdata% varies by OS version and OS language. In OS X and Linux the default location is different (don't remember at the moment). On top of that, UMS' logging is configurable, so there could be multiple log files in multiple locations. This configuration is done in logback.xml or logback.headless.xml. Because of this complicated setup, we have made the "pack debug files" function on the log tab that will pick up the log files in the configured location (yes, it will pick up multiple if  that is configured). Each configured log file will generate a separate button in the bottom of the logs tab, and hovering the mouse above the button will show the full path.

 

When it comes to the log file(s) in the "service" folder I'm unfamiliar with these, but they are probably generated by the wrapper (wrapper.exe)? Or is it generated by Windows in some way?

 

Best regards,

 

Magne Sandve

**Comment by ExSport** [comment]
E.g. LACP is switch dependant teaming but you can also make switch independent teaming easily and from Windows 2012 R2 natively without 3rd party drivers. 
On my old pc with one Gbit network card UMS shows like 10 interfaces which can be for general user no go which interface to use. UMS sees virtual ones,  ones without assigned IP etc. which should be reduced to the minimalist list. Or enable listening on all of them 0.0.0.0.
Listing virtual ones like many tunnel adapters isatap, Teredo pseudo interface, Microsoft virtual WiFi miniport adapter, VPN ones or Virtualbox ones is useless (maybe except last two:)).
Listening on all of them by default should fix many pseudo problems reported by some users on forum:))
Aaaagh it is about firewall....sorry to be little bit off topic.

**Comment by Nadahar** [comment]
@ExSport I know and agree, but as you know that's not the subject here. My point is simply that it's quite common with unconnected network interfaces - virtual or real.

**Comment by sainokawara** [comment]
Whatever you appended to my last post, exceeded my knowledge of GitHub.  I could view it but did not know how to reply.  So, I posted my response here.

It is not uncommon for people running Hyper-V to have two Ethernet adapters.  One is used to connect to the network and the other is used as the Hyper-V "Virtual Machine" adapter.  I suspect that this consititues a very small subset of users but the impact of an adapter without an IP address appears to be severe.  When the AutoIPs were eliminated, UMS functions became "lightning fast"!  I will verify that something similar cannot occur for an Ethernet adapter not being utilized for Hyper-V.

Now for the moment for which everyone has been waiting: In all of my WireShark traces and examing code for Cling/DLNA/UPnP, I have ONLY encountered UDP Port 1900.  I will keep looking to see if I can find any other Port used but, for now, that means:
1) No additional Firewall rules are needed to run UMS as a Java Application because the Java Installer creates all of the necessary Firewall rules.
2) However, to run UMS as a Windows Service, the following Firewall rule needs to be created to allow UDP Port 1900: "C:\Program Files (x86)\Java\jre1.8.0_nn\bin\java.exe".

I know that you "recommend" that users not run UMS as a Windows Service.  However, there are advantages to it.  Therefore, I would like to include the instructions in this document for the users.  Unfortunately, that does necessitate creating Firewall rules for TCP Ports 5001, 9001 and UDP Port 1900.  The average user would definitely need guidance on how to do this.  The UMS Install DOES run elevated and could be modified to just create them so they are there should the user ever decides to run UMS as a Windows Service.  Doing this introduces far less integrity exposure than the Java Installer opening ALL of the TCP/UDP Ports for Java Applications!  An alternative would be to create the rules when the users actually installs UMS as a Windows Service.  However, you just cannot do what was done for "Start automatically with Windows".  I doubt the average user would understand what to do in response to "UMS must be run as administrator in order to change this setting."  In C/C++, it is possible to request UAC elevation.  If this is possible in Java, this would be the best option.  If not, the document would have to describe how to right-click and then click on "Administrator" to run UMS elevated.  Regardless of which method is selected, the average user would probably need screen shots to walk them through the process of configuring the Windows Firewall "rules" needed.

Next, I will investigate the Firewall settings after a "clean install" and whether any Windows "components" have to be "installed" before UMS can run.  Unfortunately, I never realized that I would ever need my old XP and Vista Install DVDs and trashed them.  I can probably borrow some but are there any testers who would be able to check their systems after a "clean install" to see what, if any, actions need to be taken for UMS?  If not, I will find XP and Vista Install DVDs somewhere [but not die trying].

Hopefully, this answered most of your questions.  I will continue slogging along down this slippery slope until we have a document that, hopefully, answers every possible question in this area.

**Comment by Nadahar** [comment]
@sainokawara Great investigation, but I miss a separation of incoming and outgoing traffic.

I looked in my firewall rules now (even though it's disabled, software opens things without asking me), and it's quite shocking to see what's actually opened. I see that the common practice is to just open anything inbound and outbound, any protocol. Plex does this for example, and loads and loads of other applications. I doubt anyone would hang us for doing the same then, and it would be very simple and eliminate any firewall issues with Windows firewall. It's not "the right thing to do" though, and we need to document what to open using any firewall, not just Windows firewall. I also see rules allowing access to services there, giving access to ```svchost.exe```. There are also multiple default UPnP rules active.

As I see it, we must create rules that doesn't assume anything about existing or default rules if we are to create rules, and the documentation must do the same. You should test if creating rules for ```%SystemRoot%\system32\svchost.exe``` is working to give access to the service - that's what Windows itself uses for services. I think all services run under ```svchost.exe```.

I agree about creating rules for the service as well, and I don't actually agree about not "recommending" it - but since it requires the user to understand much more about their system and UMS, I haven't advocated changing the advice. 

If we are going to install Windows firewall rules ourselves, I think it should be handled by the installer and it should install what's needed both for application mode and service. The installer has to run elevated, so that's our shot at it. Elevating a Java application is close to impossible, you have to elevate the JVM, and you can't elevate a running process. That means you'd have to restart the JVM in elevated mode, and that would elevate any Java application. In short, elevated privileges is not practical in Java, and if such a thing were absolutely needed, a separate application handling the elevated tasks should probably be created in C/C++ and bundled.

About Vista I never had it, and I never wish to have it (as with Win 8). Windows XP however is available as a virtual machine on any Windows 7 (and probably 8 and 10) professional and above via something called "Windows virual PC XP mode". A description can be found [here] (http://windows.microsoft.com/en-us/windows7/install-and-use-windows-xp-mode-in-windows-7).

**Comment by sainokawara** [comment]
I'm glad you took a peek at what others have done.  Shocking, isn't it!  The biggest abuser was Sun/Oracle for just open ALL Ports to ALL Protocols for Java Applications.  And, we are just debating which Protocols to open for TCP Ports 5001, 9001 and UDP 1900 for Java Applications that run as a Windows Service.  Puts things in perspective, doesn't it.  I have not yet seen a need to open Port 5001 for HTTP.  However, you mentioned the early PS3 version using it for "Discovery".  I have no objection to adding it just in case it might be used in some unusual situation.  The same is be true for adding ICMP & IGMP to Port 1900 because I might have missed it in Cling or it could be added in the future even though these are definitely not in the UPnP specs.  I would rather err on the side of passing unnecessary datagrams into UMS than blocking something critical that would cause a failure observable by the user.

It appears that UMS as a Windows Service runs well with the Firewall rules posted with my TV.  [As you pointed out in your post, I didn't use the Evironment Variables.  In the future, I will use that format.]  However, I get intermittent, unexplained hangs using the Kodi Android APP.  I do not believe that it is unrelated to Firewall rules.  The one case I caught with a WireShark trace showed a "TCP Spurious Retransmission" right before the hang.  I will keep tracing because it is a good excuse for me to watch movies.  Is there another free Android UPnP Player APP, preferably on Google Play, that I could try that works with UMS?

I am glad that you agree that the UMS Installer is the right place to create the Firewall rules.  After I posted my complaint about UMS not doing a UAC elevate to change "Start automatically with Windows", I noticed that it does so for "Install/Uninstall as Windows Service".  Either my memory or powers of observation are failing me.  However, I do not believe that deferring it offers any advantage but it adds additional, unnecessary complexity to UMS which should be avoided.

Time for breakfast and coffee so I will stop temporarily for now.

**Comment by Nadahar** [comment]
In my list, most of the rules created by applications were just as that of Sun/Oracle. Allow all ports and all protocols seems to be the rule rather than the exception. It puts further weight behind my assertion that running software firewalls that nobody actually controls is quite useless, when you have to let applications configure it without the users having any idea of what they are doing. 

I can't understand why you don't see the need to open TCP 5001 incoming, because it's the primary "channel" for delivering content to renderers. Have you tried playing anything while capturing, or are you just looking at the discovery? Both TCP 5001 and 9001 must be open incoming. I don't see any need for ICMP & IGMP, to my knowledge they are not used here at all.

I haven't tried Kodi, but I've used BubbleUPnP when testing towards UMS. It doesn't support that many formats and is (as everything else on mobile) riddled with ads, but except for that it's been pretty reliable when I've used it. If you combine it with MXPlayer, it will handle a lot of formats (MXPlayer then plays what BubbleUPnP can't but still through BubbleUPnP's DLNA connection).

If UMS elevates for installing the service, that's byond our control. An external program is actually used to install/deinstall the service, so the problem remains the same - getting a Java application to elevate is close to impossible.

**Comment by sainokawara** [comment]
After a second cup of coffee, I'm back.  First, I do not know how to respond to the post by "ExPost" so let me just add, when I plugged a switch, which was connected to nothing else, into my second adapter, I got the IP that I assigned to the adapter OR, if it was to obtain the IP automatically, a 169.254.0.0/16 address.  So, these addresses documented in RFC 3927 which Microsoft refers to as "Automatic Private Internet Protocol Addressing (APIPA)" can occur but will probably be infrequent.

As for just opening Ports 5001 to ALL Protocols, I think that is overkill but the risk would be small.  However, I believe that most average users believe that Firewalls and other security products are actually "protecting" their systems.  In fact, many are willing to pay for "better" Firewalls and other such products.  I would just like UMS to be able to claim that it has the most restrictive Firewall rules of all similar products.  In answer to your question about Outbound rules, nothing is gained by using them.  Hopefully, UMS will never contain any malicious code that would be stopped by them.  And, anyone writing such code would be able to easily chuck spears virtually everywhere without impersonating UMS.  Thus, only Inbound rules need be created and documented.

I have done WireShark captures of EVERYTHING that I could within the limitations of my environment.  I was absolutely astounded at the number of TCP "recoveries" necessary.  I must have the worlds worst network [so it is good for testing]!   And, I have just upgrade to the new Version 2.0.1 so I will be doing more just to break it in.  So far, I have ONLY seen TCP Ports 5001 [, 9001 from a browser] and UDP Port 1900 all of which comply with the specs.  My only concern was those intermittent, unexplained "hangs" but I feel they are not Firewall related.

In addition to trying to diagnose the cause of the "hangs", I want to check the Firewall rules immediately after a "clean install".  I can do this for Windows 7/10.  I am working on borrowing Install DVDs for Windows XP, Vista, 8.0/8.1 so I can check these.  And, I am working on a couple of other things that may prove useful.  Well, it's 11:00, so I think I will have lunch and then get back to "work".

**Comment by Nadahar** [comment]
@sainokawara You don't have to reply to @ExSport, his point was simply that unconnected network adapters are quite common and can't be seen as a "rare case".

I've never advocated opening incoming access to port 5001 for all protocols, only for TCP. HTTP traffic is TCP and thus that's what's needed. 

But, we can't just ignore outbound rules, we need to open both outbound and inbound access for UMS. Outbound doesn't only mean to the internet, it means "out" from the computer UMS is running on to for example contact a renderer. By default Windows firewall blocks both incoming and outgoing traffic, so we need to document what's needed in both directions.

One important thing, do you have the firewall switched off when capturing with WireShark? If not, blocked packets would probably never reach WireShark.

As I said before, XP can easily be run as a virtual machine under Windows 7/10, and as to Windows Vista... who in their right mind would still be running that?

**Comment by sainokawara** [comment]
I am absolutely astounded but Windows Firewall [with Advanced Security] accepts and processes Symbolic Links like:
     %ALLUSERSPROFILE%\Oracle\Java\javapath\java.exe
In the above case, "%ALLUSERSPROFILE%" is usually "C:\ProgramData".  Since the Java Installer creates symlinks for "java.exe", "javaw.exe" and "javaws.exe", this really simplifies specifying the Wndows Firewall "rules"!

From what I have been able to find, the "recommended" way of adding/deleting rules to Windows Firewall is to use "netsh advfirewall add rule".  My knowledge of netsh and scripting is extremely limited but if you tell me which "tool" is used to build the UMS Windows Installer, I will look into what it would take to create these rules during UMS Install on Windows.  I could not find anything on the process/procedure used on GitHub.

**Comment by sainokawara** [comment]
Discovered a little known, at least to me, aspect of Java that has implications on running Java Applets, including UMS, as Windows Services:
On 64-bit Windows systems, when BOTH the 32-bit and 64-bit Java are installed, the "%ALLUSERSPROFILE%\Oracle\Java\javapath\java.exe" [and javaw.exe, javaws.exe]" symlinks point to the 64-bit Java Runtime Library, " C:\Program Files\Java\jre1.8.0_nn\bin\".  Therefore, on a 64-bit machines with both 32-bit and 64-bit Java Runtime Libraries installed, it appears that everything will run in 64-bit JVMs.  Of course, I know virtually nothing about Java.  And, if we use the Java symlinks to create the Firewall "rules", it will always be for the 64-bit Java Runtime.  And, it may be necessary to "uninstall" BOTH the 32-bit and 64-bit Java Runtime Libraries before "reinstalling" the 32-bit Java Runtime Library to "correct" these symlinks.  I will double check this and update this post after conducting this experiment.

One thing I noticed in conducting these experiments is to ALWAYS RESTART your computer IMMEDIATELY after doing an "Install/Uninstall UMS as a Windows Service" because the actual changes are not done in Windows until the machine is RESTARTED!

**Comment by Nadahar** [comment]
Great finds. I thought services normally would install without a restart, but that you had to restart to uninstall. Are you saying that that is not the case in UMS' case?

The UMS Windows installer is [NSIS] (http://nsis.sourceforge.net/Main_Page).

**Comment by sainokawara** [comment]
I have to retract using "%ALLUSERSPROFILE%\Oracle\Java\javapath\java.exe" [and javaw.exe, javaws.exe]" symlinks.  There must have been another way for the packets to "leak through" the Firewall when I did the initial testing.  But, if it did work, it was such a good approach that I am going to double check it just to be certain that it doesn't work!  Without this ability, users would have to rerun the UMS Installer after doing an uninstall/install of Java.  While rare, most users would not know to do this!  It also means having to use "...\Java\jre1.8.0_nn\bin\java.exe" when UMS creates Firewall "rules" because the Java Installer does not add any Environmental Variables!  Bummer.

I will look into NSIS.  I have heard of it but never examined it.  I also discovered that I MIGHT, and I stress the might, be able to use the aging KompoZer to create webpages.  If the HTML4 are compatible with the UMS website, even I might be able to create the necessary webpages for this little endeavor.  And, if so, is there any possibility of agreeing to "splitting the FAQs into multiple webpages"?  It is currently more than a little long and having an intermediate page to select different topics makes a lot of sense.  I hate just dropping all of this new information into that bloated webpage.

As for your question, any changes to Windows Services appear to be made at Restart time.  This makes sense because you can "delete" a Windows Service while it is running.  But, this can cause confusion if you make repeated changes using UMS "Install/Uninstall as Windows Service".  One way of avoiding this would be to tell the user that the changes do not take effect until a Restart so UMS is going to force an immediate Restart when they click "OK" like some products.  If this is too drastic an action, leave it like it is because how many people will be repeatedly changing this option without an intervening Restart?  Since it's a holiday, I have to go watch a video, err test UMS.

**Comment by Nadahar** [comment]
@sainokawara Yes, as I said, I'm used to having to restart to uninstall a service. Installing however, usually goes without. I don't know how to solve the restart issue atm, it needs further thougt.

When it comes to the FAQ, @SubJunk will have to answer that.

Regarding the firewall rules and all the different versions, that's hopeless. What we will have to do is to allow the ports in question in for all processes. 

It's not controversial imo, that's the way "normal" firewalls work anyway. The concept of process in a firewall a bit strange to me, usually a firewall is a pure network device, and there's no process information in the packages. But, Windows for obvious reasons knows about the processes too and throws that into the mix.

When it comes to outgoing ports, I tested this on a fresh Win 10 installation and there were no Java firewall rules that I could find. What happened was that, the first time I ran UMS Windows popped up with the usual "do you want to allow" (don't remember the words) dialog. When I answered yes to this, Windows created the any protocol any port rule. If this is what always happens, it means that Windows is the one to create this very "liberal" rule. Still, it might be the only way to get a outgoing allow rule created for the different Java processes. We obviously can't create an "allow anything for any process" outbound rule, so we might have to rely on Windows for the outgoing rule. It's still a problem for a service though.

**Comment by sainokawara** [comment]
"Hopeless" is a little strong.  And, yes, the first JavaApplet that runs causes Windows Firewall to create an Inbound "rule" allowing ALL Protocols & Ports for "C:\Program Files [(x86)]\Java\jre1.8.0_nn\bin\javaw.exe".  However, I believe that this is due to the decisions by "The Java Team" and not something that Windows automatically does for everything but I could be wrong.  I will try to find some non-Java program that I can test to see exactly what "rule" it creates.

However, one way or another, a user has to create two Windows Firewall "rules" for "java.exe" PRIOR to installing UMS as a Windows Service or it will not work.  If you feel that the best way of accomplishing this is to describe the process on the UMS website, possibly with screen-shots to assist users, that is an option.  Or, the UMS Installer could create automatically create the correct Firewall "rules" based upon the Java that is currently installed when it runs.  As long as the user has maintenance turned-on for Java, everything should work because I don't believe a new Java Runtime Library is created for each new "_nn" update.  But, if the user does an uninstall/install Java to apply maintenance as I do, they will have to rerun the UMS Installer OR manually update the Firewall "rules" so that they point to the new Java Runtime Library.  Neither is difficult for the user but they need to know to do it, and in the case of manually updating, what to do.

I would like to point out one thing: Although Java opens ALL Ports for ALL Protocols, it only does so for Java Applets, not everything that runs on Windows.  I do not think it would be prudent to open TCP Ports 5001, 9001 and UDP Port 1900 for ALL Programs that run on Windows.  UMS has be selective enough to only do this for Windows Services written in Java.  And, that requires the specification of the Java Runtime Library in the two Windows Firewall "rules".  So, the only decision is which of the two approaches does UMS wish to take.  If you are the "final arbitrator", it is your call.  As for me, I will implements whatever is the final decision.

**Comment by Nadahar** [comment]
I did a check on a clean Windows 10 VM I just have installed for something else. You are correct that there's two rules there allowing TCP (any port) and UDP (any port) incoming access to ```javaw.exe```.

There's no rule giving outgoing access. I do remember to have to "accept" adding a firewall rule when first starting UMS on this computer, and since the two incoming rules are the only ones Java (or UMS) related that I can find, I can't understand anything other then that those have to be the two created by Windows.

I soon figured out why there was no outgoing rule. By default, the firewall is configured as a "default block" for incoming connection and "default allow" for outgoing. As soon as I enabled "default block" (that is, block any connection not explicitly listed) UMS couldn't find any thing.

That means that the combination of the "default allow" outgoing setting and the "allow any TCP and UDP" incoming rules, the firewall is effectively disabled for TCP and UDP from UMS' POV. We've had more than enough threads on the forum where UMS starts working once they disable their firewall, so this can not always be the case. As I see it, we have to drop any assumptions about default rules and settings when creating our rules. We have to make rules that give us what we need, regardless of what choices the user has made or what defaults the OS has applied.

As I see it, the installer should ask the user if firewall rules should be installed. It should also explain that any access given to UMS is given to all Java applications, and that if the firewall rules is not installed the user will have to make sure UMS has the neccessary access him/herself. We should also give a hint to where more information can be found. The firewall rule should only apply to the "private" firewall profile, possibly also "domain". Nothing should be opened in the "public" profile.

The documentation should be general, not focusing on the Windows firewall in particular. It should say what to open, incoming and outgoing, for any firewall to allow UMS to function.

I think the only solution is to open the incoming ports for all applications in the Windows firewall rules. Those rules will apply to the service as well as any installed JVM. Neither of these ports are generally dangerous ports, there's nothing running on them by default that can be "exploited" as far as I know. Opening an incoming port is only hardful if there's an application with vulnerabilities listening on that port. At the same time, the port must be "free" (that is, no other application can be listening on it) for UMS to use it. That means that the risk of opening these are virtually non existing. 

Since DLNA/UPnP doesn't have any authentication in it and does give out information about the computer (what's in the shared folders and some other information given in headers and discovery packages) it could be argued that allowing UMS access in itself is a "security hazard". But, I don't think that's a relevant discussion, that decision is whether to allow UPnP to operate at all or not, and thus by using UMS that has implicitly been granted. Add that to the fact that in IPv4, any real world network setup that would use DLNA would have to be behind NAT that by design blocks all incoming access. The only alternative would be for someone to run public IP's on all their devices that were in the same subnet and at the same time being connected without any kind of filtering applied. If find that configuration so extremely unlikely that it's hard to take it seriously, and any person doing that would have a host of other security issues to worry about that would be much more serious than those exposed by UMS' firewall rules.

From this, we're left with outgoing access as the real problem. I still don't know what outgoing ports is used by UMS, and if allowing them for all applications would be a possibility or not. What I know for sure is that linking firewall rules to specific JVM versions is hopeless, most people use Java's automatic updater which changes this quite frequently. Getting UMS to update this dynamically isn't possible because of the elevation issues with Java. One slight possibility would be if firewall rules could be read without elevation, that would enable us to check and then to ask to user to start UMS "as an administrator" (elevated) so that UMS could update the rules. If that is possible, we could just as well do the same for incoming access. If that's not an option, I don't really see a good way of solving outgoing access.

When it comes to the final decision, there's nobody appointed to that. We have to come up with a suggested solution, and then we'd have to present that to the others. They can all read this discussion though, but my impression is that most of them are allergic to reading long texts so they have probably stopped following this issue a long time ago.

Don't bother with the HTML, the important bit is the text itself in whatever form. If it's going to be integrated in the FAQ, @SubJunk will have to handle inserting the text there anyway.

**Comment by sainokawara** [comment]
Because I worked with system code rather than user code, I had to go through this futile exercise just to satisfy myself that there was no easy solution.  If there had been a simple solution, I would have said that is the way to do it.  But, the Windows Firewall does not resolve symlinks and appears to only support Microsoft defined System Environment Variables.  So, after a sleepless night, I arrived at almost the identical conclusion.

I believe that the only option for UMS is to just open Ports 5001 and 9001 to TCP, and, to open Port 1900 to UDP.  As you have so often stated, the risk is minuscule and the consequences are insignificant.  About all a nefarious program could discover is the titles of the music and videos being played.  UDP Port 1900 is more problematic because, theoretically, it would be possible to actually reconstruct the songs or video.  However, this would require an immense amount of data that, as you point out, would have to be shipped through the Router's Firewall!  For anyone who wants to completely eliminate this risk the UMS documentation could describe, with screen shots, exactly how to update the two UMS Firewall "rules" so only UMS can listen on these three Ports [by these "rules"].  But, the user would be responsible for manually updating them when the Java Runtime Libraries are "updated".  The symptoms would be that UMS running as a Windows Service simply stops working.  The corrective action is almost trivial for someone who has added the Program/Service option to a Firewall "rule".

Believe it or not, there is no need for any Firewall Outbound "rules" in the UMS situation for security!  As you discovered, it would be a total nightmare determining what Ports & Protocols UMS uses.  And, then there are the "Dynamic or Private Ports" (49152-65535) which are automatically open so a process like UMS and its associated processes can communicate with each other.

Although I would like the UMS documentation to describe the process for manually creating the necessary Firewall "rules", especially for adding the more restrictive Program/Service option, it may be possible for UMS to create/delete these simply and easily with a minimum of code and a small test impact.  The code that "Installs/Uninstalls UMS as a Windows Service" runs elevated.  It should be very easy to add the Microsoft recommeded "netsh advfirewall firewall ... " commands to create and delete the two "rules".  Of course, these would be the simple open TCP Ports 5001, 9001 and UDP Port 1900.  If this were done, nearly all of the users would be able to run UMS as a Windows Service but would not even have to know about the Firewall let alone its "rules".

Don't bother looking it up but if you happen to know which routine contains the "Install/Uninstall UMS as a Windows Service", please let me know.  If you don't, I will eventially find it.  I think we are getting there, where ever there is.

By the way, I would suggest that XP users are on their own and have to do everything manually.  Because of several disasters, Microsoft completely reworked the XP Windows Firewall and produced Windows Firewall with Advanced Security for Vista and follow on releases.  Every Microsoft document you read on Windows Firewall has all kinds of "weasle words" concerning XP support.  It would probably be best for UMS to avoid any issues in this area.

**Comment by sainokawara** [comment]
I believe I have located the section of code in UMS that would needs to be changed.  Give me a little while to investigate further.

Here are the "netsh" commands that would have to be issued to create and delete the UMS Firewall "rules".  It, of course, has to run elevated.  I have tested this on Windows 10 and it works exactly as it should.  I have no reason to suspect that it won't work on Vista.  The only relevant limitation of which I am aware is that port ranges must be added in a different manner but none are needed for UMS.

To add the UMS Firewall "rules", do the following when UMS is being "Installed as a Windows Service.  The leading _ are only to force alignment on this forum and should not be specified:

___netsh   advfirewall firewall
______delete rule     name="UMS TCP Ports"
___netsh   advfirewall firewall
______delete rule     name="UMS UDP Ports"

___netsh   advfirewall firewall
______add rule     name="UMS TCP Ports"
_________localport=5001,9001   protocol=tcp   dir=in   action=allow
___netsh   advfirewall firewall
______add rule     name="UMS UDP Ports"
_________localport=1900   protocol=udp   dir=in   action=allow

When UMS is being "Uninstalled as a Windows Service", it is only necessary to do the first two "netsh" commands to delete the Firewall "rules".  And, when UMS is "Installed as a Windows Service", the user should NOT be able to change the two TCP Port Numbers!  I don't know how easy it is but I like the Microsoft approach of just graying them out to indicate that they cannot be changed but displaying their values.  If the user want to change these Ports, they must "Uninstall as a Windows Service", change their values and then "Install as a Windows Service".  It really is this simple!

Oh, and if the code to elevate could be wrapped around changing "Start automatically with Windows" instead of producing an Error Message, it would really be nice!  But, this is not required for Firewall "rules", just something that I almost consider a bug.

**Comment by Nadahar** [comment]
>  For anyone who wants to completely eliminate this risk the UMS documentation could describe, with screen shots, exactly how to update the two UMS Firewall "rules" so only UMS can listen on these three Ports [by these "rules"].

That wouldn't really help security as UMS itself delivers information to anyone who asks. Also, UMS is probably the only application on that installation using those ports, otherwise there would be a conflict if both were started. For anyone that really want to minimize the risk, they should explicitly define the other party (the renderer(s)) that the given access rules applies to, but I think that's byond the scope of this effort.

> Believe it or not, there is no need for any Firewall Outbound "rules" in the UMS situation for security! As you discovered, it would be a total nightmare determining what Ports & Protocols UMS uses. And, then there are the "Dynamic or Private Ports" (49152-65535) which are automatically open so a process like UMS and its associated processes can communicate with each other.

I disagree. The documentation shouldn't assume anything about defaults in the firewall, it shouldn't assume that the firewall in question is Windows firewall and it shouldn't assume that the OS is Windows. If UMS needs a broad range of outgoing ports, the range (from which they are picked) should be determined and stated as a requirement. When it comes to the Windows firewall rules I might be inclined to accept that we don't specify outgoing rules since there's no way to tie it to the correct JVM/Java executable. UMS doesn't use TCP/IP to communicate with associated processes, it uses pipes.

I don't see the need for a documentation for the Windows firewall in particular. The Windows firewall rules are simple and straight forward for anyone that knows what a firewall is, so a general documentation of the ports and protocols needed should suffice. Anyone that doesn't understand a firewall should stick with the automatically created rules.

>  The code that "Installs/Uninstalls UMS as a Windows Service" runs elevated.

Yes, but this is not Java. This is a 3. party executable for Windows, probably written in C/C++, that do have the ability to elevate. 

About Windows XP, the installer should detect this and then inform them to read the (general) firewall documentation or turn the firewall off. I agree that we don't go through the hazzle to automate that.

>  And, when UMS is "Installed as a Windows Service", the user should NOT be able to change the two TCP Port Numbers!

That's not possible, all configuration is in plain text files and configuration can be done either there or in the GUI. AFAIR the ports aren't configurable from the GUI, but only in the configuration files. A comment could be inserted in the configuration file stating that if you change these and use Windows firewall, the automatically generated firewall rules won't apply.

> Oh, and if the code to elevate could be wrapped around changing "Start automatically with Windows" instead of producing an Error Message, it would really be nice! But, this is not required for Firewall "rules", just something that I almost consider a bug.

It's off topic yes, but as stated before, elevation in Java is so impractical that we just as might consider it impossible. My idea is instead to change this so that it only acts on the current user. That means it wouldn't require elevated privileges to change it. It's on my "to do list".

**Comment by sainokawara** [comment]
I should have included a "profile=private" in the "netsh" command to limit connections to only "private" and not "public" or "domain" networks.  Also, there is a 'netsh advfirewall firewall show rule name="UMS TCP Ports" ' that does not require elevation that would allow UMS to obtain the Port information from Windows Firewall.

I am limiting my "security" strictly to whether a program operating outside of UMS could obtain information that legitimate UMS users may not want revealed.  Whether UMS is "secure" is an entirely separate issue.  Having been involved in such never ending quests, I would suggest that it not be pursued.

Again, there is no need for any Outbound "rules".  And, if this were attempted, bugs which caused UMS to fail would be found many years from now.  It would be virtually impossible to even identify all of the "system services" used by UMS, all of its subcomponents (Cling, SSDP, UPnP, wrapper.exe, etc.) and all of their subcomponents in every possible circumstance.

UMS utilizes a function of "wrapper.exe" to "Install as a Windows Service".  I do not know whether the "wrapper.commandfile" function or any of its other functions has the ability to request elevation before issuing a command?  The need is there so I am surprised that it wouldn't.  There are Open Source C routines which permit issuing commands in an elevated state.  If these cannot be used, it appears that the only possible placement for the "netsh" commands would be in the UMS Installer.  Currently, there is no method of specifying the Ports in this process.  If this were the choice, I would suggest that the user have to manually enter the "netsh" commands to predefine whatever Ports they want to use.  And, these could be specified to UMS as they are today OR UMS could retrieve the information that the user created for Windows Firewall by using the "netsh advfirewall firewall show rule ... " command which does not require elevation.

Whatever technique is used, however, precludes the user being able to change the Ports via the existing UMS method.  If they did, the Firewall "rules" would not be updated and UMS would not function when run as a Windows Service.  So, the user would be responsible for coordinating the changes and manually updating Windows Firewall via "netsh" commands.

**Comment by sainokawara** [comment]
Ignoring the problem with restricting the Firewall rules so they only apply just to UMS or, at most, only to Java applets; in an attempt to answer your question on what OutBound Ports must be “opened”, I conducted several “experiments”.  In the following, I have BLOCKED all OutBound Ports 256 or greater except TCP Port 5001 and UDP Port1900.  UMS actually runs but:
–→ Client Discovery of UMS can be excruciatingly slow.
–→ Client operations are “inconsistent and non-repeatable”
And, when an error does occur for the TV client, the TV is usually unable to recover so responds by just “Disconnecting from UMS”.  The Kodi APP on my Android Tablet, apparently, has more robust recovery and UMS operation appears to continue normally even though I see error messages in the Logs.

The developers probably have better things to do than work on such a problem but I believe there is a bug lurking here so am sending this to document it.  However, **I am most interested in what appear to be socket Ports: 211, 1255, 1452 and 1460.**

Wrapper.log of problem which occurred at 10:42:59:
INFO  10:38:16.113 [WrapperSimpleAppMain] Using address /192.168.1.32 found on network interface: name:eth1 (Realtek PCIe GBE Family Controller)
INFO  10:38:16.113 [WrapperSimpleAppMain] Created socket: /192.168.1.32:5001
INFO  10:38:16.160 [WrapperSimpleAppMain] WEB interface is available at: http://192.168.1.32:9001
INFO  10:38:16.238 [WrapperSimpleAppMain] Using the following UUID configured in UMS.conf: a464ecc2-672a-4175-ba23-09912988635e
INFO  10:38:16.499 [WrapperSimpleAppMain] The server is now available for renderers to find
INFO  10:38:16.889 [UPNPHelper] Access granted to /192.168.1.32 by rule: 192.168.1.32
INFO  10:39:34.936 [UPNPHelper] Access granted to /192.168.1.20 by rule: 192.168.1.20
INFO  10:41:58.839 [New I/O worker #5] Checking shared folder: N:\Music
INFO  10:41:58.839 [New I/O worker #5] Checking shared folder: N:\Videos

**<Observe that an error was successful recovered or circumvented>
DEBUG 2016-01-04 10:42:59.570 [pool-11-thread-2] Starting C:\Program Files (x86)\Universal Media Server\win32\mplayer.exe -identify -endpos 0 -v -ao null -vc null -vo null -dvd-device N:\Videos\480 Video Title, Set 1, DISC 1\VIDEO_TS dvd://1**
INFO  10:42:59.586 [pool-11-thread-1] Exception in thread "pool-11-thread-1" java.lang.NullPointerException
INFO  10:42:59.586 [pool-11-thread-1] 	at net.pms.dlna.DVDISOTitle.resolveOnce(DVDISOTitle.java:211)
INFO  10:42:59.586 [pool-11-thread-1] 	at net.pms.dlna.DLNAResource.resolve(DLNAResource.java:1460)
INFO  10:42:59.586 [pool-11-thread-1] 	at net.pms.dlna.DLNAResource.syncResolve(DLNAResource.java:1452)
INFO  10:42:59.586 [pool-11-thread-1] 	at net.pms.dlna.DLNAResource.run(DLNAResource.java:1255)
INFO  10:42:59.586 [pool-11-thread-1] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
INFO  10:42:59.586 [pool-11-thread-1] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
INFO  10:42:59.586 [pool-11-thread-1] 	at java.lang.Thread.run(Unknown Source)
**<Could [pool-11-thread-2] have triggered this problem?>**

<The video played on the Kodi Android APP appeared to work correctly>
INFO  10:43:32.847 [StartPlaying Event] Started playing Title 1 on your Kodi
INFO  10:44:00.066 [StopPlaying Event] Stopped playing Title 1 on your Kodi
<Kodi APP on Android TABLET terminated>

**Comment by Nadahar** [comment]
@sainokawara There definitly is a bug there, it seems like ```DVDISOTitle.media``` is ```null```. However, we'd need [full debug files] (http://www.universalmediaserver.com/forum/viewtopic.php?f=9&t=556) to stand a chance at figuring out why. We should try to keep this issue separated from the firewall issue, as I doubt they are related.

When you say the ports seem like socket ports, what does that mean? Are you referring to source or destination port numbers?

**Comment by sainokawara** [comment]
The following may have to be adjusted based upon my limited understanding of Java:

The good news is that Cling does provide a constructor to change the TCP Port on which it listens:
___UpnpService upnpService =
______new UpnpServiceImpl(new DefaultUpnpServiceConfiguration(###));
as described in Section 3.1 of the Cling Documentation.

The bad news is that it would require a UMS external change!  So, you probably don't want to do it?

However, without this change, Cling requests any available TCP Port so it is completely unpredictable what Port it will use.  Since it is not unreasonable to state in any UMS Firewall document that ALL Dynamic/Ephemeral/Private TCP/UDP must not be BLOCKED, this is probably a "non-problem" anyway.  However, I just wanted everyone to know if this limitation.

**Comment by Nadahar** [comment]
@sainokawara I'm not sure I understand what you mean by external change. It's easy to specify a port number [there] (https://github.com/UniversalMediaServer/UniversalMediaServer/blob/master/src/main/java/net/pms/network/UPNPControl.java#L282) - but it would require some handling if the port was already in use. Still shouldn't be too bad, I think that's a useful option. I'm assuming that this will replace the formerly mentioned UDP port 1900?

**Comment by sainokawara** [comment]
This is not the UDP Port 1900.  This is a TCP Port used by Cling.  A specific Port number can be selected as I mentioned in my previous post but there is no existing UMS external to specify it.  As long as UMS is happy with it assuming almost any and every possible value [limited by the Dynamic/Ephemeral/Private Port range], I see little need for such a specification.  It, of course, must not be blocked but that is really true for this entire range.

**Comment by Nadahar** [comment]
I don't follow you now, this is a listening port - that is an incoming port. Wouldn't it be smart to specify that, so our incoming rules are 5001, 9001, xxx TCP, 1900 UDP?

**Comment by sainokawara** [comment]
I don't disagree [a double negative!].  But, if the following works, it might be solution that would not require an external changes: UMS could force Cling to use TCP Port 7001 which would almost always work.  Hopefully, if that Port is unavailable, Cling would then just continue using any available in the Dynamic/Private Port range.

I might need a day or two to answer all of the questions concerning this Port.  And, I have quite a few!

**Comment by Nadahar** [comment]
I doubt Cling would "fall back" to a dynamic port if e.g 7001 is in use. I looked at the code, and when no port argument is given, "0" i sent. That argument probably ends up directly at socket creation in the underlying OS. But, that's ok - we just have to find a way to detect if it is blocked so that we can inform the user.

**Comment by sainokawara** [comment]
This is more mysterious than I imagined.  Apparently, in the UPnP specifications, there is a method for a process such as Cling to have UPnP Devices send it "information" [I don't know exactly what or under what circumstances, YET] directly to a TCP [Don't know about UDP] Port.  When I brought up UMS, Cling got a TCP Port in the Private range.  I only traced for a one or so but my TV sent Cling three messages to Cling!  All three were just 46 bytes of data and just looking at them, they appeared identical.  I will see what I can find out exactly why and what is going on.  Would you prefer that I just update this post or create a new post?

**Comment by Nadahar** [comment]
This has gotten too long to go back to check for modifications, just add new or corrected information.

**Comment by sainokawara** [comment]
I always seem to learn something new everyday:
Even when TCP Ports 256 and greater were BLOCKED both Inbound and Outbound by the Firewall,
a socket was created, used and would, apparently, work for Local communications.  However, WireShark confirms that there was absolutely no traffic from/to TCP Port 9986 across the Adapter!  So, there appears to be no way to even be able to detect that the Firewall is BLOCKING remote traffic.  Now, my only concern is when the specific TCP Port cannot be used because another process is using it.  This should be extremely rare but I want to investigate exactly what Cling will do when directed to use a TCP Port that is in-use. If Cling continues to function, I suspect that there is enough other sources, specifically UDP Port 1900, that UMS would continue to function, possibly, with slightly degraded UPnP Discovery.  If so, UMS could simply force Cling to use TCP Port 7001.  I hope I have learned enough Java to be able to "implement" this so I can test it.

Since Cling is part of the UMS package, I really think this TCP Port 7001 needs to be included in the Port document.  But, I just can't justify a new UMS external only for this purpose.


| Jan 07, 2016 5:49:19 PM org.fourthline.cling.UpnpServiceImpl <init>
| INFO: >>> Starting UPnP service...
| Jan 07, 2016 5:49:19 PM org.fourthline.cling.UpnpServiceImpl <init>
| INFO: Using configuration: net.pms.network.UPNPControl$1
| Jan 07, 2016 5:49:19 PM org.fourthline.cling.transport.RouterImpl <init>
| INFO: Creating Router: org.fourthline.cling.transport.RouterImpl
| Jan 07, 2016 5:49:19 PM org.fourthline.cling.transport.impl.MulticastReceiverImpl init
| INFO: Creating wildcard socket (for receiving multicast datagrams) on port: 1900
| Jan 07, 2016 5:49:19 PM org.fourthline.cling.transport.impl.MulticastReceiverImpl init
| INFO: Joining multicast group: /239.255.255.250:1900 on network interface: Realtek PCIe
| Jan 07, 2016 5:49:19 PM org.fourthline.cling.transport.impl.StreamServerImpl init
| INFO: Created server (for receiving TCP streams) on: /192.168.1.32:**9986**
| Jan 07, 2016 5:49:19 PM org.fourthline.cling.transport.impl.DatagramIOImpl init
| INFO: Creating bound socket (for datagram input/output) on: /192.168.1.32
| Jan 07, 2016 5:49:19 PM org.fourthline.cling.UpnpServiceImpl <init>
| INFO: <<< UPnP service started successfully
| Jan 07, 2016 5:49:19 PM org.fourthline.cling.protocol.RetrieveRemoteDescriptors describe
| **WARNING: Device descriptor retrieval failed, no response: http://192.168.1.17:7676/smp_18_**

**Comment by sainokawara** [comment]
Even with UMS 5.5.0, "Resume Video" failed.  I believe this is due to the fact that my TV has difficulty recovering from errors.  I will retest this using the Kodi APP on my Android Tablet which handles error recovery much better.

Remember, I can only see what is written to the UMS Logs and WireShark only records packets which traverse an Adapter.  Even so, it seems that Cling's TCP Port adds little, if anything, to UMS.  During one test run, WireShark only captured activity for about one-tenth of a second!  For anyone who wishes to know, here are the specifics:

___1) TV ->  Cling = NOTIFY "RenderingControl"
___2) Cling -> TV = OK
___3) TV -> Cling = NOTIFY "AVTransport"
___4) Cling -> TV = OK
___5) TV -> Cling = NOTIFY "ConnectionManager"
___6) Cling -> TV = OK

During this interval, there was also a fine selection of other Datagrams to/from Cling's TCP whose data content added little to my understanding of what was occurring.  That was the ONLY activity during the entire test.  I will run some longer and more complex tests of UMS to see whether something/anything more interesting occurs.  If not, why is there any concern about this TCP Port?

**Comment by Nadahar** [comment]
That information is needed, it's where the renderer informs Cling of it's capabilities.

**Comment by sainokawara** [comment]
Any information it receives/sends is redundant.  Cling obtains everything it needs from UDP Port 1900 and UMS.  UMS obtains everything it needs from its TCP Port [and, if it listens, UDP Port 1900].

I tested everything I could including "volume control" and "resume video" using a Kodi APP on my Android Tablet with the Cling TCP Port "disabled" and everything worked correctly!  I will retest with another WireShark trace but I am extremely confident that the Cling TCP Port is not required for Cling or UMS to function properly.  I am not suggesting removing the specification of the default port, "0", in
___"DefaultUpnpServiceConfiguration sc = new DefaultUpnpServiceConfiguration()"
just changing it to the UMS TCP Port number + 1.  This port will almost aways be available but if it is not, it just doesn't matter!  This greatly simplifies documenting the Ports required for UMS operation.

And, here is the Log showing that Cling's TCP Port really was not used in these tests:

_ org.fourthline.cling.UpnpServiceImpl <init>
_INFO: >>> Starting UPnP service...
_org.fourthline.cling.UpnpServiceImpl <init>
_INFO: Using configuration: net.pms.network.UPNPControl$1
_org.fourthline.cling.transport.RouterImpl <init>
_INFO: Creating Router: org.fourthline.cling.transport.RouterImpl
_org.fourthline.cling.transport.impl.MulticastReceiverImpl init
_INFO: Creating wildcard socket (for receiving multicast datagrams) on port: 1900
_org.fourthline.cling.transport.impl.MulticastReceiverImpl init
_INFO: Joining multicast group: /239.255.255.250:1900 on network interface:
_org.fourthline.cling.transport.RouterImpl startAddressBasedTransports
_**WARNING: Failed to init StreamServer: java.net.BindException: Address already in use: bind**
_org.fourthline.cling.transport.RouterImpl startAddressBasedTransports
_WARNING: Removing unusable address: /192.168.1.32
_org.fourthline.cling.transport.RouterImpl handleStartFailure
_**INFO: Unable to initialize network router, no network found.**
_org.fourthline.cling.UpnpServiceImpl <init>
_INFO: <<< UPnP service started successfully

**Comment by Nadahar** [comment]
I'm not convinced by the tests, the question is if cling use that info to build a "device" profile. If it does, UMS will need that for detecting renderers in some cases, even if your renderers are recognized without this information. Why did you get a BindException? Did you set it to the same port that UMS uses?

**Comment by sainokawara** [comment]
I set the Cling TCP Port to 135 which is used by a Windows function.  When the Cling TCP Port is "disabled", it cannot be used to send/receive anything.

As indicated in the Log of my prior post by,
___**"INFO: Creating wildcard socket (for receiving multicast datagrams) on port: 1900"**,
Cling is using UDP Port 1900 and obtaining everything it needs from it.  Remember, on Windows [but I don't know whether there is an equivalent Linux function], SSDP/UPnP are operating in parallel and are also using UDP Port 1900.

On longer and more complex tests, WireShark traces show two different, distinct patterns:

For the TV, when the Cling Port is OPEN, every second there is a:
1) Cling -> TV = POST "/smp_26_   Soapaction: GetPositionInfo"
2) TV -> Cling = OK
These are exchanged using Ephemeral TCP Ports, not the Cling TCP Port!
These do NOT occur when the Cling Port is "disabled".

For the Kodi APP on my Android Tablet when the Cling Port is OPEN, every second:
1) Cling (TCP Port) -> APP (Port#1) = NOTIFY "AVTransport"
__and shortly thereafter APP (Port#1) -> Cling (TCP Port) = OK
2) Cling (TCP Port) -> APP (Port#2) =  NOTIFY "AVTransport"
__and shortly thereafter APP (Port#2) -> Cling (TCP Port) = OK
These do NOT occur when the Cling Port is "disabled".

To repeat, "disabling the Cling TCP Port" does not appear to affect UMS functions in anyway **on Windows**.  Perhaps, this is not true on Linux, I don't know.  Since I am not advocating removing the Cling TCP Port, just changing the Port Number assignment mechanism to simplify the Firewall documentation, is it really necessary to test on Linux?  Another option would be to add a new UMS external to permit the specification of a particular Cling TCP Port.  Or, as I have previously stated, it is just as easy for a user to use the current specification for the UMS Port Number to specify two adjecent TCP Port Numbers.

But, I am just beating a dead horse.  Whoever makes such decisions just needs to make the decision.

**Comment by Nadahar** [comment]
My primary concern isn't what portnumber to assign, but what role the port has. To know for sure, I think I'd have to look at the UPnP port 1900 traffic. 

I don't understand UDP port 1900 - if it's a listening port, how can both "UMS" and Cling bind to it? Or are there other rules for multicast?

**Comment by sainokawara** [comment]
Most peoples' confusion arises because everyone talking about TCP/UDP and Ports leave out one very important piece of information: They are associated with an "IP Address"!  There are three basic "IP Addresses" about which we are concerned in my test environment:

1) 127.0.0.1 or the "Localhost"
2) The IP Addresses of my test devices:
__192.168.1.17 or the IP Address of my TV
__192.168.1.20 or the IP Address of my Tablet
__192.168.1.32 or the IP Address of my Windows 10 system to the outside world
3) The "IP Address" 239.255.255.250 or the "Multicast Group for SSDP/UPnP"

You are probably familiar with the usual sequence to create a communication link between two points by doing a:

__1) socket_create
__2) socket_bind to specify the "IP Address" AND Port on which to listen
__3) socket_listen to just wait around for a socket_connect OR
____socket_connect to connect to another listening socket

However, there was a need for applications to be able to communicate with many other applications possibly on different systems so "Multicasting Groups" were created.  The two best known "IP Addresses" for these are the IP range 224.0.0.1-x and 239.255.255.250 which was assigned to SSDP/UPnP.  If an application creates a socket and binds it to one of these "IP Addresses", it can communicate with all of the other applications who have similiarly created a socket and bound to the same "IP Address".  So, all UPnP Devices whether Clients or Servers [including Cling for UMS] create a socket and bind to "IP Address" 239.255.255.250 AND UDP Port 1900.

And, the UPnP specifications have architected a long list of "requests" and the corresponding "responses" that UPnP Devices must implement for this "Multicasing IP Address" and their actual "IP Addresses".  Among these is a "Discovery Process" that allows for the interchange of all kinds of UPnP Device "descriptive information" and actual "IP Addresses" including the TCP Port upon which it is listening for UPnP "commands" using this special UDP Port 1900.

Does this explain why so many can be listening and responding on the same Port?

**Comment by Nadahar** [comment]
I know very well that a port is related to an IP address, that's not the point. What I know is that in unicast two applications cannot bind to the same port on the same IP address at once. I don't really know in detail how multicast works, and that's probably the problem here. I do understand that the data is sent to all devices listening to a given port, but that doesn't (to me) explain how two sockets listening to the same address/port on the same device can exist. But it probably can, and that's probably the answer to this. That also gives that there can be no two-way communication over multicast, there's no way to direct a response to a particular "member", and that's probably why any "broadcast" message sent on port 1900 contains a "reply address" that is unicast. The discussed TCP 5002/7001 is probably Cling's "reply channel" that any multicast member will contact for one-on-one communication, just like TCP 5001 is for UMS.

Does this make sense to you?

**Comment by sainokawara** [comment]
Exactly, for "Multicast", any number of sockets can listen/broadcast so both Cling and UMS could have a UDP Port 1900 socket.  Because any message is sent to all listeners, they should be limited to "Discovery" related activities and the massive data transfer operations should be done with standard point-to-point ehemeral or private sockets.

On another note, I tested with the Cling TCP Port "disabled"  while the Windows Services "SSDP Discovery" and "UPnP Device Host" were completely "Disabled" and not running.  UMS worked exactly as it should!  This is probably identical to a UNIX environment for which no Packages exists, to my knowledge, that provide these equivalent functions.  I am confident that UMS would function correctly in a UNIX environment even if the Cling TCP Port 5002 was unavailable.  At the moment, I am not interested in setting up a Java IDE which would allow running these test in a true UNIX environment.  And, I really don't believe it is necessary.

I have a few more things I want to check and then I will post a summary of all of the relevant points for the Firewall document.  The only open issue would be the decision by the developers on whether the existing UMS TCP Port specification can be used to also specify Cling's TCP Port (add ONE) OR whether a new external must be added to allow any Port Number.  And, if the latter, I would suggest that the change be made by "Edit UMS configuration file manually" rather than adding a new line/box to the "General Configuration" page.  Keep in mind, "UMS appears to run fine without Cling's TCP Port!"

**Comment by sainokawara** [comment]
I would propose the following for the UMS Firewall Documentation.  On a Windows system, it is easiest for a user to add them using the "netsh advfirewall firewall" command.  And, it is easy to "translate" them into the format required for another Firewall.  The leading character on the lines under the "netsh advfirewall firewall" command lines are only for alignment and should not be entered.

netsh   advfirewall firewall
  * delete rule     name="UMS Inbound TCP Ports"

netsh   advfirewall firewall
  * delete rule     name="UMS Inbound UDP Ports"

netsh   advfirewall firewall
  * add rule     name="UMS Inbound TCP Ports"
    * localport=5001,5002,9001   protocol=tcp   dir=in   action=allow   profile=private

netsh   advfirewall firewall
  * add rule     name="UMS Inbound UDP Ports"
    *localport=1900   protocol=udp   dir=in   action=allow   profile=private

If you are unwilling to accept the default and just leave ALL of the TCP/UDP Outbound ports open, these ports must explicitly be open in the Outbound direction:

netsh   advfirewall firewall
  * delete rule     name="UMS Outbound TCP Ports"

netsh   advfirewall firewall
  * delete rule     name="UMS Outbound UDP Ports"

netsh   advfirewall firewall
  * add rule     name="UMS Outbound TCP Ports"
    * localport=5001,5002,9001   protocol=tcp   dir=out   action=allow

netsh   advfirewall firewall
  * add rule     name="UMS Outbound UDP Ports"
    *localport=1900   protocol=udp   dir=out   action=allow

Finally, ALL of the Dynamic/Ephemeral/Private Ports should be open in the Outbound direction which is the default.  This Port range varies by platform but the IANA assigned range is 49152-65535.

This is the complete list for UMS [including all of its components/subcomponents].

**Comment by Nadahar** [comment]
I see no reason to have documentation for each firewall - and the list is long. The documentation only need to specify what direction/protocol/port needs to be accessible for UMS.

Those that do not understand firewalls should use the (by UMS) created firewall rules.

**Comment by sainokawara** [comment]
I just did the research.  You are free to accept/reject it in whole or part.  And, how it is presented is completely up to the person creating the webpage.  However, each keyword in the command is relevant to the Firewall options that specify exactly what the Firewall allows/blocks and are required to properly document the ports used by UMS.

**Comment by sainokawara** [comment]
Just documentation if/when the UMS Installer ever creates Windows Firewall "rules":

It is trivial to create the UMS Windows Firewalls "rules" when generating a Windows Installer with NSIS.  Perhaps, there is another approach that is more sophisticated or consistent with UMS but the following has been tested and does work.  The " \" [space backslash] at the end of the two lines indicate "line continuation" to NSIS and should not be included if the two lines are merged into a single line.

1) These must run with "RequestExecutionLevel admin"

2) Feel free to change the "ExecToLog" to "Exec" if it is more consistent with UMS.  Add the following six lines to the UMS Install section:

     nsExec::ExecToLog 'netsh advfirewall firewall delete rule name="UMS TCP Ports"'
     nsExec::ExecToLog 'netsh advfirewall firewall delete rule name="UMS UDP Ports"'

     nsExec::ExecToLog 'netsh advfirewall firewall add rule name="UMS TCP Ports" \
         protocol=tcp localport=5001,5002,9001 dir=in action=allow profile=private'
     nsExec::ExecToLog 'netsh advfirewall firewall add rule name="UMS UDP Ports" \
         protocol=udp localport=1900 dir=in action=allow profile=private'

3) Add the following two lines to the UMS Uninstall section:

     nsExec::ExecToLog 'netsh advfirewall firewall delete rule name="UMS TCP Ports"'
     nsExec::ExecToLog 'netsh advfirewall firewall delete rule name="UMS UDP Ports"'

4) If you change the rule names, they must unambiguously reflect that they belong to UMS!  And, do NOT include anything that could create confusion with another Function, Product or Project especially not "SSDP" or "UPnP"!  I am partial to "UMS TCP Ports" and "UMS UDP Ports" because they are short but descriptive.

5) Please, PLEASE include the two DELETES in BOTH the Install and Uninstall sections so there is no rubbish left behind if UMS is uninstalled [unlike most Products/Projects] and multiple, possibly conflicting, "rules" aren't created when the UMS Installer is run multiple times!

**Comment by Nadahar** [comment]
@sainokawara This is useful, I agree about deleting them both during unstall and uninstall. I also noticed that you've sneaked in 5002 :wink: 

I think we have to figure out how to make a dialog in the installer to ask if we should install the firewall rules. Uninstalling them should be done without asking.

**Comment by sainokawara** [comment]
Once I learned enough NSIS to be dangerous, it was fairly easy to use.  Personally, I would just always add the "Firewall Rules" since the risk is so low.  However, if you really feel that it is necessary to provide a user option, here is what I would do:  Add a Name to the existing "Section" and a new "Section" so the following is displayed on the Components Page:

   [Checkbox] "UMS Base"
   [Checkbox] "Program  Files"
   [Checkbox] "Firewall Rules"
   [Checkbox] "Start Menu Shortcuts"

I actually got this to work but was unable to upload the "Snipping Tool" file for everyone to admire?

**Comment by Nadahar** [comment]
@sainokawara Maybe you can tell me how to quickly "build" a test NSIS installer for testing small modifications to the scripts? I'm sure I can find out, but if you already know...

**Comment by sainokawara** [comment]
I have completed coding & testing, including running in "silent mode", the changes necessary to control the UMS Install, including "AviSynth", from the newly displayed Installer "Components" page.  I also discovered that the UMS 5.5.0 Installer does not properly check to guarantee that UMS is not running!  I "corrected" this but this change is not "guaranteed" because I encountered an "NSIS Plugin" problem that prevented me from completing all of the testing.

Specificly, I was unable to get the Installer to compile because I got the following error:
___File: "LockedList64.dll"->"$PLUGINSDIR\LockedList64.dll" 9728 bytes
___**Plugin not found, cannot call LockedList::AddModule**
Interestingly, 9,728 bytes is the length of "LockedList64.dll"!  So, I don't understand why the plugin wasn't found for the next statement?  I understand that there should be a .nsh file for it.  From what I found on the web, some claim that I only need "MUI.nsh" which is there.  But, there was no .nsh file when I downloaded this Plugin from the NSIS website?

I encountered a similiar problem with for the "AccessControl" and "nsSCM" Plugins.  Perhaps, someone more knowledgeable about NSIS Plugins than I can tell me what I have to do so that these three Plugins are found?  I will try to attach the "almost" finished code so that anyone who knows the solution can test it if I am able to circumvent the GitHub "errors" preventing me from upload it.  However, I would really like to know the answer to this mystery!

**Comment by sainokawara** [comment]
The UMS code and NSIS installer script that will set Cling's TCP port to UMS's TCP port number plus one has been completed.  Also, there was no validity checking of the TCP port range so a restriction was added limiting UMS to TCP ports 1024 - 65534.  When an invalid TCP port is specified in the .CONF file, an error is logged and the default UMS TCP port, 5001, is used.  From the "GeneralTab", the change is ignored until a valid port is entered.  Altering this last behavior would be difficult with the current way data is entered/processed.

The NSIS Windows installer script was also changed to use a "Components Page".  And, it was changed, as was the "Windows Menu", to simplify the script and be more consistent with other products/projects with which I am familiar.  I would have attached a .PNG file illustrating the change but I have never been able to attach a file successfully at GitHub.  Although it looks different than the existing install, it is functionally identical.  If AviSynth is already installed at the correct level, the files are not copied even though it is checked.  The installer default is to always creating the UMS rules for the Windows Firewall with Advanced Security necessary to run UMS as a Windows Service using the default TCP ports, 5001, 5002 and 9001; and UDP port 1900.  Although it is not difficult for a knowledgeable user to change these port assignments, it would be helpful if there were documentation on exactly how to do this.  Also, a minor bug was fixed in the installer script which prevented the UMS installer from detecting UMS running in a 32-bit JVM on a 64-bit system.

I was told that "SubJunk" would update the documentation.  Although I was a developer of system level software, not a writer, I am willing to take a stab at it.  I do not know whether it would generate the correct level of HTML but I did find "KompoZer" which seems simple enough for me to actually use.  As I have indicated, in addition to documenting the ports used by UMS, UMS should describe how to detect a port conflict and how to change the ports in Windows Firewall as well as in UMS.

From my testing on Windows, it appears UMS functions is not affected by a conflict on Cling's TCP port.  However, this may be due to the fact that Microsoft added UPnP/SSDP support to Windows and UMS is unaffected because it obtains all of the necessary information via its TCP/UDP ports.  Currently, UNIX/Linux testing is extremely difficult on my system.  I hope to remedy this soon.  However, I believe that it is necessary to determine/document whether a conflict on Cling's TCP port would affect UMS function on UNIX/Linux systems.  Although the likelihood a conflict is extremely small, one possible "solution", if it does, would be to continue allowing Cling to obtain and use a "dynamic/private" TCP port.  Unfortunately, this would prevent UMS from documenting a specific TCP port assignment and users would have to keep this port range "open".  I suspect virtually all UMS users do this anyway.

Finally, I would appreciate knowing whether there is any interest in me pursing this or other matters further.  If not, as a UMS user, I will simply contribute monetarily to UMS and look elsewhere for other forms of contributions that I could make.  Thank you,

**Comment by Nadahar** [comment]
This stalled from the lack of interest.

**Comment by polygirl** [comment]
@Nadahar - Add an exception to Windows Firewall on install was implemented so closing this issue.
